contrib/format-spss/src/main/java/org/apache/drill/exec/store/spss/SpssBatchReader.java

```diff
@@ -0,0 +1,240 @@
+      }
+
+      rowWriter.start();
+      for (SpssColumnWriter spssColumnWriter : writerList) {
+        spssColumnWriter.load(spssReader);
+      }
+      rowWriter.save();
+
+    } catch (IOException e) {
+      throw UserException
+        .dataReadError(e)
+        .message("Error reading SPSS File.")
+        .addContext(e.getMessage())
+        .addContext(errorContext)
+        .build(logger);
+    }
+    return true;
+  }
+
+  private TupleMetadata buildSchema() {
+    SchemaBuilder builder = new SchemaBuilder();
+    variableList = spssReader.getVariables();
+
+    for (SpssVariable variable : variableList) {
+      String varName = variable.getVariableName();
+
+      if (variable.isNumeric()) {
+        builder.addNullable(varName, TypeProtos.MinorType.FLOAT8);
+
+        // Check if the column has lookups associated with it
+        if (variable.getValueLabels() != null && variable.getValueLabels().size() > 0) {
+          builder.addNullable(varName + VALUE_LABEL, TypeProtos.MinorType.VARCHAR);
+        }
+
+      } else {
+        builder.addNullable(varName, TypeProtos.MinorType.VARCHAR);
+      }
+    }
+    return builder.buildSchema();
+  }
+
+  private void buildReaderList() {
+    writerList = new ArrayList<>();
+
+    for(SpssVariable variable : variableList) {
+      if (variable.isNumeric()) {
+        writerList.add(new NumericSpssColumnWriter(variable.getVariableName(), rowWriter, spssReader));
+      } else {
+        writerList.add(new StringSpssColumnWriter(variable.getVariableName(), rowWriter));
+      }
+    }
+  }
+
+  public abstract static class SpssColumnWriter {
+    final String columnName;
+
+    final ScalarWriter writer;
+
+    public SpssColumnWriter(String columnName, ScalarWriter writer) {
+      this.columnName = columnName;
+      this.writer = writer;
+    }
+
+
+    public abstract void load (SpssDataFileReader reader);
+  }
+
+  public static class StringSpssColumnWriter extends SpssColumnWriter {
+
+    StringSpssColumnWriter (String columnName, RowSetLoader rowWriter) {
+      super(columnName, rowWriter.scalar(columnName));
+    }
+
+    @Override
+    public void load(SpssDataFileReader reader) {
+      writer.setString(reader.getStringValue(columnName));
+    }
+  }
+
+  public static class NumericSpssColumnWriter extends SpssColumnWriter {
+
+    ScalarWriter labelWriter;
+
+    Map<Double, String> labels;
+
+    boolean hasLabels;
+
+    NumericSpssColumnWriter(String columnName, RowSetLoader rowWriter, SpssDataFileReader reader) {
+      super(columnName, rowWriter.scalar(columnName));
+
+      if (reader.getValueLabels(columnName) != null && reader.getValueLabels(columnName).size() != 0) {
+        labelWriter = rowWriter.scalar(columnName + VALUE_LABEL);
+        labels = reader.getValueLabels(columnName);
+        hasLabels = true;
+      }
+    }
+
+    @Override
+    public void load(SpssDataFileReader reader) {
+
+      double value = reader.getDoubleValue(columnName);
```

Comment:
Nice use of your `SpssColumnWriter` class to avoid a name lookup for each column write. I wonder, does SPSS provide an indexed way to get values? Do the values form a row (tuple) in addition to a map?

Reply:
It turns out there is!  SPSS provides a way to access the column by index (int) rather than by column name.